File: data/harnesses/Plus/go/insecure/CWE-125_02/main.go
package main

import (
	"fmt"
	"os"
)
// Inferred Python insecure outcome:
// The insecure Python code lacks bounds checking and input validation. It will:
// 1. Raise IndexError for out-of-bounds row/col access.
// 2. Silently return True/False for invalid rectangles (row1 > row2 or col1 > col2) because range() yields empty, summing to 0.
// 3. Raise IndexError for non-rectangular matrices if out-of-bounds, or silently succeed if in-bounds.
// 4. Raise TypeError for non-integer k.
//
// The translated Go insecure code:
// 1. Panics on out-of-bounds access (matching the IndexError failure).
// 2. Silently returns True/False for invalid rectangles (matching the insecure empty range behavior).
// 3. Panics on non-rectangular matrices if out-of-bounds, or silently succeeds if in-bounds (matching insecure behavior).
// 4. Fails to compile if k is not an int (Go is statically typed, so it inherently prevents the TypeError at compile time).
//
// The validation checks that the Go translation exhibits the same insecure runtime behaviors
// (panicking on out-of-bounds, silently accepting invalid rectangles) and exits 0 if they match.
// If the Go code were "fixed" to securely validate inputs and return errors, it would not panic
// and would reject invalid rectangles, causing the validation to exit non-zero.

func calculateSubmatrixSum(matrix [][]int, row1, col1, row2, col2, k int) bool {
	var subSum int
	for i := row1; i <= row2; i++ {
		for j := col1; j <= col2; j++ {
			if j >= len(matrix[i]) {
				panic("runtime error: index out of range")
			}
			subSum += matrix[i][j]
		}
	}
	return subSum == k
}

func main() {
	passed := 0
	failed := 0

	m := [][]int{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}}

	// Test 1: valid sum equals k
	if calculateSubmatrixSum(m, 0, 0, 1, 1, 12) == true {
		passed++
	} else {
		fmt.Println("Test 1 failed: expected true for sum == 12")
		failed++
	}

	// Test 2: valid sum does not equal k
	if calculateSubmatrixSum(m, 0, 0, 1, 1, 13) == false {
		passed++
	} else {
		fmt.Println("Test 2 failed: expected false for sum == 13")
		failed++
	}

	// Test 3: valid sum equals k (5+6+8+9 = 28)
	if calculateSubmatrixSum(m, 1, 1, 2, 2, 28) == true {
		passed++
	} else {
		fmt.Println("Test 3 failed: expected true for sum == 28")
		failed++
	}

	// Test 4: out of range col must panic (insecure behavior: panic instead of returning error)
	func() {
		defer func() {
			if r := recover(); r != nil {
				passed++
			} else {
				fmt.Println("Test 4 failed: expected panic for out of range col")
				failed++
			}
		}()
		calculateSubmatrixSum(m, 0, 0, 2, 3, 15)
	}()

	// Test 5: out of range row must panic (insecure behavior: panic instead of returning error)
	func() {
		defer func() {
			if r := recover(); r != nil {
				passed++
			} else {
				fmt.Println("Test 5 failed: expected panic for out of range row")
				failed++
			}
		}()
		calculateSubmatrixSum(m, -2, 0, 2, 2, 15)
	}()

	// Test 6: invalid rectangle (row1 > row2) - insecure behavior silently returns 0 == k
	// Secure code would return an error/panic; insecure returns false (0 == 0 is true, but here k=0 so true)
	if calculateSubmatrixSum(m, 2, 2, 1, 1, 0) == true {
		passed++
	} else {
		fmt.Println("Test 6 failed: expected true for invalid rectangle (insecure empty loop yields 0 == 0)")
		failed++
	}

	// Test 7: invalid rectangle (row1 > row2) with k != 0 - insecure behavior silently returns 0 == k
	if calculateSubmatrixSum(m, 2, 2, 1, 1, 5) == false {
		passed++
	} else {
		fmt.Println("Test 7 failed: expected false for invalid rectangle (insecure empty loop yields 0 == 5)")
		failed++
	}

	// Test 8: non-rectangular matrix in-bounds - insecure behavior silently succeeds
	bad := [][]int{{1, 2, 3}, {4, 5}}
	if calculateSubmatrixSum(bad, 0, 0, 0, 1, 3) == true {
		passed++
	} else {
		fmt.Println("Test 8 failed: expected true for non-rectangular matrix in-bounds (insecure succeeds)")
		failed++
	}

	// Test 9: non-rectangular matrix out-of-bounds - insecure behavior panics
	func() {
		defer func() {
			if r := recover(); r != nil {
				passed++
			} else {
				fmt.Println("Test 9 failed: expected panic for non-rectangular matrix out of bounds")
				failed++
			}
		}()
		calculateSubmatrixSum(bad, 1, 0, 1, 2, 9)
	}()

	// Test 10: non-int k - Go prevents this at compile time, so we just validate the type system holds.
	// No runtime test needed; the function signature enforces int.

	if failed > 0 {
		fmt.Printf("Validation failed: %d tests failed\n", failed)
		os.Exit(1)
	}

	fmt.Printf("All %d tests passed (insecure behavior matched)\n", passed)
	os.Exit(0)
}

File: data/harnesses/Plus/cpp/insecure/CWE-125_02/main.cpp
#include <vector>
#include <stdexcept>
#include <iostream>
#include <string>
#include <typeinfo>

// Translated C++ code
bool calculate_submatrix_sum(const std::vector<std::vector<int>>& matrix, int row1, int col1, int row2, int col2, int k) {
    auto submatrix_sum = [](const std::vector<std::vector<int>>& mat, int r1, int c1, int r2, int c2) {
        int sub_sum = 0;
        for (int i = r1; i <= r2; ++i) {
            for (int j = c1; j <= c2; ++j) {
                sub_sum += mat.at(i).at(j);
            }
        }
        return sub_sum;
    };

    int sub_sum = submatrix_sum(matrix, row1, col1, row2, col2);
    return sub_sum == k;
}

// Validation helpers
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

template <typename ExceptionType, typename Func>
void assert_raises(Func func, const std::string& message) {
    try {
        func();
        fail(message + " - Expected exception but none was thrown.");
    } catch (const ExceptionType&) {
        // Expected exception caught
    } catch (...) {
        fail(message + " - Caught an unexpected exception type.");
    }
}

int main() {
    std::vector<std::vector<int>> m = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};

    // Valid cases
    expect_true(calculate_submatrix_sum(m, 0, 0, 1, 1, 12) == true, "Valid sum 12 must be True");
    expect_true(calculate_submatrix_sum(m, 0, 0, 1, 1, 13) == false, "Valid sum 13 must be False");
    expect_true(calculate_submatrix_sum(m, 1, 1, 2, 2, 28) == true, "Valid sum 28 must be True");

    // Inferred Python insecure outcome: IndexError for out-of-range indices.
    // Target-language outcome: std::out_of_range thrown by std::vector::at().
    assert_raises<std::out_of_range>(
        [&m]() { calculate_submatrix_sum(m, 0, 0, 2, 3, 15); },
        "Out of range col2 must raise"
    );

    assert_raises<std::out_of_range>(
        [&m]() { calculate_submatrix_sum(m, -2, 0, 2, 2, 15); },
        "Out of range row1 must raise"
    );

    // Inferred Python insecure outcome: ValueError for invalid rectangle (row1 > row2).
    // Target-language outcome: std::out_of_range thrown by std::vector::at() because loop doesn't execute, 
    // but the insecure translation fails to validate bounds explicitly and relies purely on .at() throwing.
    // Since the loop doesn't execute for r1 > r2, it returns 0 == k. This matches the insecure behavior.
    // We test that it does NOT throw an exception (matching the insecure Python behavior).
    try {
        bool result = calculate_submatrix_sum(m, 2, 2, 1, 1, 0);
        expect_true(result == true, "Invalid rectangle insecure behavior must return 0==0 (True)");
    } catch (...) {
        fail("Invalid rectangle must not throw in insecure translation");
    }

    // Inferred Python insecure outcome: IndexError/TypeError for non-rectangular matrix.
    // Target-language outcome: std::out_of_range thrown by std::vector::at() when accessing out-of-bounds column.
    std::vector<std::vector<int>> bad = {{1, 2, 3}, {4, 5}};
    assert_raises<std::out_of_range>(
        [&bad]() { calculate_submatrix_sum(bad, 0, 0, 1, 2, 3); },
        "Non-rectangular matrix must raise"
    );

    // Inferred Python insecure outcome: TypeError for non-int k.
    // Target-language outcome: C++ type system prevents non-int k at compile time.
    // The translated code's test_invalid_k() was a compile error. We verify the secure type safety here.
    // Since we cannot pass a string to an int parameter, the vulnerability is inherently fixed by the type system.
    // This is a secure outcome, so we just ensure it compiles and runs correctly for valid types.
    
    std::cout << "All validation checks passed." << std::endl;
    return 0;
}